### PR TITLE
Query for local beatmap in multiplayer room listing

### PR DESCRIPTION
**Pull request by nyquillerium**
- Closes https://github.com/ppy/osu/issues/4401

### CODE REVIEW

File: osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs [hunk 1]
```diff
@@ -46,17 +46,19 @@ protected override DelayedLoadWrapper CreateDelayedLoadWrapper(Drawable content,
         protected override Drawable CreateDrawable(BeatmapInfo model)
         {
             Drawable drawable;
+            BeatmapInfo localInfo = new BeatmapInfo();
 
-            var localBeatmap = beatmaps.GetWorkingBeatmap(model);
+            if (model != null)
+                localInfo = beatmaps.QueryBeatmap(p => p.OnlineBeatmapID == model.OnlineBeatmapID);
```
Comment (peppy) [inline osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs]: this should only be performed if the first online lookup fails.

File: osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs [hunk 2]
```diff
@@ -46,17 +46,19 @@ protected override DelayedLoadWrapper CreateDelayedLoadWrapper(Drawable content,
         protected override Drawable CreateDrawable(BeatmapInfo model)
         {
             Drawable drawable;
+            BeatmapInfo localInfo = new BeatmapInfo();
```
Comment (peppy) [inline osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs]: initialisation here is unnecessary.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs (hunk 2 above)
Comment (nyquillerium) [reply osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs]: wont compile unless I initialize here

Comment (peppy) [reply osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs]: use null?

**Comment by nyquillerium** [comment]
Discovered some additional problems with matchsongselect IDs. Blocking while I investigate

**Comment by nyquillerium** [comment]
separate PR will be made for the match song select background stack

File: osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs [hunk 3]
```diff
@@ -46,17 +46,21 @@ protected override DelayedLoadWrapper CreateDelayedLoadWrapper(Drawable content,
         protected override Drawable CreateDrawable(BeatmapInfo model)
         {
             Drawable drawable;
+            BeatmapInfo localInfo;
 
-            var localBeatmap = beatmaps.GetWorkingBeatmap(model);
+            if (model != null && model.ID == 0)
```
Comment (peppy) [inline osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs]:
```suggestion
            if (model?.ID == 0)
```

File: osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs [hunk 4]
```diff
@@ -46,17 +46,21 @@ protected override DelayedLoadWrapper CreateDelayedLoadWrapper(Drawable content,
         protected override Drawable CreateDrawable(BeatmapInfo model)
         {
             Drawable drawable;
+            BeatmapInfo localInfo;
```
Comment (peppy) [inline osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs]: no need for this variable.

File: osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs [hunk 5]
```diff
@@ -46,17 +46,21 @@ protected override DelayedLoadWrapper CreateDelayedLoadWrapper(Drawable content,
...
 
-            var localBeatmap = beatmaps.GetWorkingBeatmap(model);
+            if (model != null && model.ID == 0)
+                localInfo = beatmaps.QueryBeatmap(p => p.OnlineBeatmapID == model.OnlineBeatmapID);
```
Comment (peppy) [inline osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs]: this is going to catastrophically fail if `model.OnlineBeatmapID` is null. it'll probably return a random beatmap.

File: osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs [hunk 6]
```diff
@@ -46,17 +46,21 @@ protected override DelayedLoadWrapper CreateDelayedLoadWrapper(Drawable content,
...
+            else
+                localInfo = model;
```
Comment (peppy) [inline osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs]: unnecessary, remove `localInfo` variable

File: osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs [hunk 7]
```diff
@@ -46,17 +46,21 @@ protected override DelayedLoadWrapper CreateDelayedLoadWrapper(Drawable content,
...
 
             if (model?.BeatmapSet?.OnlineInfo != null)
             {
                 drawable = new BeatmapSetCover(model.BeatmapSet, beatmapSetCoverType);
             }
-            else if (localBeatmap.BeatmapInfo.ID != 0)
+            else if (localInfo?.ID != null)
```
Comment (peppy) [inline osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs]: did you mean `> 0`?

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs [hunk 8]
```diff
@@ -47,16 +47,17 @@ protected override Drawable CreateDrawable(BeatmapInfo model)
         {
             Drawable drawable;
 
-            var localBeatmap = beatmaps.GetWorkingBeatmap(model);
+            if (model?.OnlineBeatmapID != null)
```
Comment (peppy) [inline osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs]: isn't this in the wrong place? shouldn't it be inside the `else if`? 

99% sure this is going to regress behaviour as the database doesn't provide `OnlineInfo` and you are now forcing usage of the database beatmap.

Comment (nyquillerium) [reply osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs]: We have to query for the beatmap here though, or else we won't know whether or not there is an actual local version (the original issue we had). This is why I had the localinfo variable earlier.

Comment (nyquillerium) [reply osu.Game/Beatmaps/Drawables/UpdateableBeatmapBackgroundSprite.cs]: Nevermind that, I can find a different solution

**Comment by peppy** [comment]
This is more wrong now. You can't possibly be testing this.

**Comment by peppy** [comment]
Closing in favour of #4454.

**Comment by nyquillerium** [comment]
Closing in favor of https://github.com/ppy/osu/pull/4454

**Pull request closed without merging**